pyximport/_pyximport3.py

```diff
@@ -205,283 +213,142 @@ def build_module(name, pyxfilename, pyxbuild_dir=None, inplace=False, language_l
-                 inplace=False, language_level=None):
-        _debug("PyxLoader created for loading %s from %s (init path: %s)",
-               fullname, path, init_path)
-        self.fullname = fullname
-        self.path, self.init_path = path, init_path
-        self.pyxbuild_dir = pyxbuild_dir
-        self.inplace = inplace
-        self.language_level = language_level
+    def create_module(self, spec):
+        try:
+            so_path = build_module(spec.name, pyxfilename=spec.origin, pyxbuild_dir=self._pyxbuild_dir,
+                                   inplace=self._inplace, language_level=self._language_level)
+            self.path = so_path
+            spec.origin = so_path
+            return super().create_module(spec)
+        except Exception as failure_exc:
+            _debug("Failed to load extension module: %r" % failure_exc)
+            if pyxargs.load_py_module_on_import_failure and spec.origin.endswith(PY_EXT):
+                spec = importlib.util.spec_from_file_location(spec.name, spec.origin,
+                                                              loader=SourceFileLoader(spec.name, spec.origin))
+                mod = importlib.util.module_from_spec(spec)
+                assert mod.__file__ in (spec.origin, spec.origin + 'c', spec.origin + 'o'), (mod.__file__, spec.origin)
+                return mod
+            else:
+                tb = sys.exc_info()[2]
+                import traceback
+                exc = ImportError("Building module %s failed: %s" % (
+                    spec.name, traceback.format_exception_only(*sys.exc_info()[:2])))
+                raise exc.with_traceback(tb)
 
-    def load_module(self, fullname):
-        assert self.fullname == fullname, (
-            "invalid module, expected %s, got %s" % (
-            self.fullname, fullname))
-        if self.init_path:
-            # package
-            #print "PACKAGE", fullname
-            module = load_module(fullname, self.init_path,
-                                 self.pyxbuild_dir, is_package=True,
-                                 build_inplace=self.inplace,
-                                 language_level=self.language_level)
-            module.__path__ = [self.path]
-        else:
-            #print "MODULE", fullname
-            module = load_module(fullname, self.path,
-                                 self.pyxbuild_dir,
-                                 build_inplace=self.inplace,
-                                 language_level=self.language_level)
-        return module
+    def exec_module(self, module):
+        try:
+            return super().exec_module(module)
+        except Exception as failure_exc:
+            import traceback
+            _debug("Failed to load extension module: %r" % failure_exc)
+            raise ImportError("Executing module %s failed %s" % (
+                    module.__file__, traceback.format_exception_only(*sys.exc_info()[:2])))
```

Comment:
Do we have tests for these failure cases?

Reply:
Unfortunately not. How this failure can be implemented? `exec_module()` internally uses low level machinery which I cannot "break":
https://github.com/python/cpython/blob/6217864fe5f6855f59d608733ce83fd4466e1b8c/Lib/importlib/_bootstrap_external.py#L1236-L1240
Corrupting .so file causes segmentation fault. Is it possible to use mock for this test?